corehq/apps/export/export.py

```diff
@@ -109,6 +112,164 @@ def path(self):
         return self._path
 
 
+class _PaginatedWriter(object):
+
+    def __init__(self, writer, page_length=None):
+        self.format = writer.format
+        self._path = None
+        self.page_length = page_length or MAX_EXPORTABLE_ROWS
+        self.pages = Counter()
+        self.rows_written = Counter()
+        # An instance of a couchexport.ExportWriter
+        self.writer = writer
+        self.file_handle = None
+
+    @contextlib.contextmanager
+    def open(self, export_instances):
+        """
+        Open the _PaginatedWriter for writing. This must be called before using _PaginatedWriter.write().
+        A _PaginatedWriter can only be opened once.
+        """
+
+        assert self._path is None
+
+        self.name = self._get_name(export_instances)
+        self.headers = self._get_headers(export_instances)
+        self.table_names = self._get_table_names(export_instances)
+
+        # Create and open a temp file
+        fd, self._path = tempfile.mkstemp()
+        with os.fdopen(fd, 'wb') as file_handle:
+            self.writer.open(
+                self._get_paginated_headers().iteritems(),
+                file_handle,
+                table_titles=self._get_paginated_table_titles(),
+                archive_basepath=self.name
+            )
+            yield
```

Comment:
You probably want to wrap this `yield` in `try`/`finally`. Otherwise, if there's an exception raised inside the context, the next line (`self.writer.close()`) will not be called.